wcfsetup/install/files/lib/system/service/worker/Encryption.class.php

```diff
@@ -0,0 +1,187 @@
+            $salt
+        );
+        // Section 4
+        $payload = Encryption::addPadding($payload, $serviceWorker->contentEncoding);
+
+        $tag = '';
+        $encryptedText = \openssl_encrypt(
+            $payload,
+            'aes-128-gcm',
+            $cek,
+            OPENSSL_RAW_DATA,
+            $nonce,
+            $tag
+        );
+
+        if ($serviceWorker->contentEncoding === ServiceWorker::CONTENT_ENCODING_AESGCM) {
+            $headers['Encryption'] = 'salt=' . Base64UrlSafe::encode($salt);
+            $headers['Crypto-Key'] = 'dh=' . Base64UrlSafe::encode($newPublicKey);
+        }
+
+        return Encryption::getEncryptionContentCodingHeader(
+            $serviceWorker->contentEncoding,
+            $salt,
+            $newPublicKey
+        ) . $encryptedText . $tag;
+    }
+
+    private static function getSharedSecret(JWK $publicKey, #[\SensitiveParameter] JWK $privateKey): string
+    {
+        $result = \openssl_pkey_derive(
+            ECKey::convertPublicKeyToPEM($publicKey),
+            ECKey::convertPrivateKeyToPEM($privateKey),
+            256
+        );
+        \assert($result);
+        return \str_pad($result, 32, "\0", STR_PAD_LEFT);
+    }
+
+    private static function addPadding(string $payload, string $contentEncoding): string
+    {
+        $length = Binary::safeStrlen($payload);
+        $paddingLength = ServiceWorkerHandler::MAX_PAYLOAD_LENGTH - $length;
+
+        if ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AES128GCM) {
+            return \str_pad($payload . "\2", $paddingLength + $length, "\0", STR_PAD_RIGHT);
+        } elseif ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AESGCM) {
+            return \pack('n*', $paddingLength) . \str_pad($payload, $paddingLength + $length, "\0", STR_PAD_LEFT);
+        } else {
+            throw new \RuntimeException('Unknown content encoding "' . $contentEncoding . '"');
+        }
+    }
+
+    private static function createInfo(string $type, ?string $context, string $contentEncoding)
+    {
+        if ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AESGCM) {
+            \assert($context !== null);
+            \assert(Binary::safeStrlen($context) === 135);
+
+            return 'Content-Encoding: ' . $type . "\0" . Encryption::CURVE_ALGORITHM . $context;
+        } elseif ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AES128GCM) {
+            return 'Content-Encoding: ' . $type . "\0";
+        } else {
+            throw new \RuntimeException('Unknown content encoding "' . $contentEncoding . '"');
+        }
+    }
+
+    private static function getEncryptionContentCodingHeader(
+        string $contentEncoding,
+        string $salt,
+        string $publicKey
+    ): string {
+        if ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AES128GCM) {
+            return $salt
+                . \pack('N*', 4096)
+                . \pack('C*', Binary::safeStrlen($publicKey))
```

Comment:
These two `pack` calls and especially the second one look fishy to me.

Reply:
But this is exactly what is needed.
> salt (16) | rs (4) | idlen (1) | keyid (idlen)

https://datatracker.ietf.org/doc/html/rfc8188#section-2.1

Reply:
I see. The asterisk was confusing to me (because there is only a single parameter) and I expected more than a single byte for the length. I'll make a suggestion to make this easier to follow.